src/main/scala/sigmastate/eval/TreeBuilding.scala

```diff
@@ -106,29 +106,52 @@ trait TreeBuilding extends RuntimeCosting { IR: Evaluation =>
     }
   }
 
-  def buildValue(mainG: PGraph, env: DefEnv, s: Sym, defId: Int): SValue = {
+  def buildValue(mainG: PGraph,
+                 env: DefEnv,
+                 s: Sym,
+                 defId: Int,
+                 constantsProcessing: Option[ConstantsProcessing]): SValue = {
     import builder._
-    def recurse[T <: SType](s: Sym) = buildValue(mainG, env, s, defId).asValue[T]
-    object In { def unapply(s: Sym): Option[SValue] = Some(buildValue(mainG, env, s, defId)) }
+    def recurse[T <: SType](s: Sym) = buildValue(mainG, env, s, defId, constantsProcessing).asValue[T]
+    object In { def unapply(s: Sym): Option[SValue] = Some(buildValue(mainG, env, s, defId, constantsProcessing)) }
     s match {
       case _ if env.contains(s) =>
         val (id, tpe) = env(s)
         ValUse(id, tpe) // recursion base
       case Def(Lambda(lam, _, x, y)) =>
         val varId = defId + 1       // arguments are treated as ValDefs and occupy id space
       val env1 = env + (x -> (varId, elemToSType(x.elem)))
-        val block = processAstGraph(mainG, env1, lam, varId + 1)
+        val block = processAstGraph(mainG, env1, lam, varId + 1, constantsProcessing)
         val rhs = mkFuncValue(Vector((varId, elemToSType(x.elem))), block)
         rhs
       case Def(Apply(fSym, xSym, _)) =>
         val Seq(f, x) = Seq(fSym, xSym).map(recurse)
         builder.mkApply(f, IndexedSeq(x))
       case Def(th @ ThunkDef(root, _)) =>
-        val block = processAstGraph(mainG, env, th, defId)
+        val block = processAstGraph(mainG, env, th, defId, constantsProcessing)
         block
       case Def(Const(x)) =>
         val tpe = elemToSType(s.elem)
-        mkConstant[tpe.type](x.asInstanceOf[tpe.WrappedType], tpe)
+        constantsProcessing match {
+          case Some(s: ExtractConstants) =>
+            val constant = mkConstant[tpe.type](x.asInstanceOf[tpe.WrappedType], tpe)
+              .asInstanceOf[ConstantNode[SType]]
+            s.extract(constant)(builder)
+          case Some(_: InjectConstants) =>
```

Comment:
@greenhat this looks like overcomplication. Note that we may have both constants to extract and also Placeholders from previous layers

Reply:
Fixed.